application.go

```diff
@@ -0,0 +1,51 @@
+package sdk
+
+import (
+	"context"
+
+	"github.com/cosmos/gogoproto/grpc"
+	"github.com/pokt-network/poktroll/x/application/types"
+)
+
+// ApplicationClient is an ApplicationClient implementation that uses the gRPC
+// query client of the application module.
+// QueryClient is made public because it should eventually become an interface, as it is being consumed here.
+//
+//	More details in the following link: https://go.dev/wiki/CodeReviewComments#interfaces
+type ApplicationClient struct {
+	types.QueryClient
+}
+
+// NewApplicationClient creates a new application client with the provided gRPC connection.
+func NewApplicationClient(grpcConn grpc.ClientConn) *ApplicationClient {
+	return &ApplicationClient{
+		QueryClient: types.NewQueryClient(grpcConn),
+	}
+}
+
+// GetAllApplications returns all applications in the network.
+func (ac *ApplicationClient) GetAllApplications(
+	ctx context.Context,
+) ([]types.Application, error) {
+	req := &types.QueryAllApplicationsRequest{}
+	res, err := ac.QueryClient.AllApplications(ctx, req)
+	if err != nil {
+		return []types.Application{}, err
+	}
+
+	return res.Applications, nil
+}
+
+// GetApplication returns the details of the application with the given address.
+func (ac *ApplicationClient) GetApplication(
+	ctx context.Context,
+	appAddress string,
+) (types.Application, error) {
+	req := &types.QueryGetApplicationRequest{Address: appAddress}
+	res, err := ac.QueryClient.Application(ctx, req)
+	if err != nil {
+		return types.Application{}, err
+	}
+
+	return res.Application, nil
+}
```

Comment:
Shouldn't this file be under `clients/application.go` now? A bit confused

Reply:
This is the default implementation, placed at the top level (and will be used by the Gateway). The previous one is kept as a custom client (no strong opinions on the latter: we could keep it as a place holder for future potential custom clients, or we could scrap it).

Reply:
Going back to https://github.com/pokt-network/shannon-sdk/pull/8#issuecomment-2161524918, my understanding was:

`SDK` contains:
1. Interfaces
2. Default implementation

Users of `SDK` repo can either:
1. Implement their own client (following the interfaces)
2. Use the default implementation

Having the `SDK` define both of the following is confusing:
1. Default implementation
2. Custom implementation

My suggestion is:
1. `clients` define the default implementation
2. Portal (our gateway) use the default implementation
3. Others can either use it as well or do their own thing

Reply:
OK SGTM and is basically what this PR is trying to do. Having a single client to avoid confusion is a good call. If at some point there is a need for more/custom clients, we can discuss the best approach.

One suggested change from me:
All the clients/sdk should live at the top-level package, with no sub-packages (at-least for now), i.e. the top directory of the repo will look like this (with top-level package called `sdk`):
```
application.go
block.go
relay.go
```

Portal/others can use the SDK by simply importing 1 package:
```
import (
     ....

     sdk "github.com/pokt-network/shannon-sdk"
)

...
appClient := sdk.ApplicationClient{
   ....
}
```

So if the following makes sense, I'll go ahead and update the PR:

- The updated `application.go` at the top level.
- No other Application Client functionality anywhere else in the repo to avoid confusion.
- The PR doesn't touch any other files (e.g. no changes to `example-clients` in this PR)

If this sounds good, then the next PRs will continue this path:
- Break a single client functionality out of existing code (e.g. `RelayClient`, `BlockClient`, etc.)
- Refactor and move to the top-level package `sdk`.
- Continue updating the gateway code to use the new clients.